chore(deadline): document correct usage of secrets management admin credentials (#592)
aws/aws-rfdk: packages/aws-rfdk/lib/deadline/README.md

### Configuring Deadline Secrets Management on the Repository

> When using Deadline 10.1.19 or higher, RFDK enables [Deadline Secrets Management](https://docs.thinkboxsoftware.com/products/deadline/10.1/1_User%20Manual/manual/secrets-management/deadline-secrets-management.html) on the `Repository` by default. RFDK is able to create administator credentials for Deadline Secrets Management, store them in AWS Secrets Manager, and configure the Deadline Repository with those credentials. When RFDK generates the admin credentials in a Secret for you, it sets the Secret's removal policy to `RETAIN` so that you will not lose your credentials if you destroy the CloudFormation Stack that they are in. Even with the removal policy set to `RETAIN`, destroying the Stack that the Secret is in could orphan it from your application and cause a new Secret to be created if the Stack is redeployed. To prevent this, RFDK will put the Secret in the same Stack as your database so that their lifecycles are in sync.

Reviewer: That is good but I was thinking about warning users that even after farm destroying they will still have a Secret.
It could be a problem if they do development and deploy it several times or using it for automatically testing. 
So they have to change property in Secret Managements configuration or destroy it manually.
What do you think?

Author: Yeah good point, I've reworked this section a bit more.

### Configuring Deadline Secrets Management on the Repository

When using Deadline 10.1.19 or higher, RFDK enables [Deadline Secrets Management](https://docs.thinkboxsoftware.com/products/deadline/10.1/1_User%20Manual/manual/secrets-management/deadline-secrets-management.html) on the `Repository` by default. To perform administrator functions for Deadline Secrets Management, credentials need to be created. You can either create your own in an AWS Secrets Manager Secret and have it imported into your RFDK app, or have RFDK generate the credentials and create the Secret for you.

> When RFDK generates the administrator credentials, it sets the Secret's removal policy to `RETAIN` so that you will not lose them if you destroy the CloudFormation Stack that they are in. If you do want to delete this Secret, you will have to destroy it manually or change the `credentialsRemovalPolicy` to `DESTROY`. If you destroy the Stack your Secret is located with the policy set to `RETAIN` your existing Secret will be orphaned from your RFDK application and redeploying the Stack will generate a new one. To help prevent orphaning a Secret that contains administrator credentials that are still in use, RFDK places it in the Stack that contains your database, so that their lifecycles are in sync.

Reviewer 2: ```suggestion
When RFDK generates the administrator credentials, it sets the Secret's removal policy to `RETAIN` so that you will not lose them if you destroy the CloudFormation Stack that they are in. If you do want to delete this Secret, you will have to destroy it manually or change the `credentialsRemovalPolicy` to `DESTROY`. If you destroy the Stack that contains your Secret and its policy set to `RETAIN`, your existing Secret will be orphaned from your RFDK application and re-deploying the Stack will generate a new one. To help prevent orphaning a Secret that contains administrator credentials that are still in use, RFDK places it in the Stack that contains your database, so that their lifecycles are in sync.
```

Author: Updated

> If your database is imported into your RFDK app rather than being generated by it, RFDK will not allow your app to have the credentials generated automatically and you'll need to create your own Secret with the credentials you'd like to use and import it. If you would like to use your own credentials for Deadline Secrets Management, you can do so by storing them in AWS Secrets Manager and providing them to the `Repository` construct. The Secret must be a JSON document with the following format:

Reviewer 2: ```suggestion
If your database is imported into your RFDK app rather than being generated by it, RFDK will not generate credentials during deployment. You must create your own Secret with the credentials you'd like to use and [import it](https://docs.aws.amazon.com/cdk/api/latest/docs/aws-secretsmanager-readme.html#importing-secrets).

If you would like to use your own credentials for Deadline Secrets Management, you can do so by storing them in AWS Secrets Manager and providing them to the `Repository` construct. The Secret must be a JSON document with the following format:
```

Author: Updated

> ```json

Reviewer 2: ```suggestion
```jsonc
```

Author: Fixed

### Configuring Deadline Secrets Management on the Repository

By default, [Deadline Secrets Management](https://docs.thinkboxsoftware.com/products/deadline/10.1/1_User%20Manual/manual/secrets-management/deadline-secrets-management.html) is enabled on the `Repository`.

> when using Deadline 10.1.19 or higher. RFDK is able to create administator credentials for Deadline Secrets Management, store them in AWS Secrets Manager, and configure the Deadline Repository with those credentials. When RFDK generates the admin credentials for you, it puts them in the same Stack as your database so that their lifecycles are in sync and you cannot accidentally delete the credentials that the database is configured to use. If your database is imported into your RFDK app rather than being generated by it, RFDK will not allow your app to have credentials generated automatically and you'll need to create your own AWS Secret with the credentials you'd like to use and import it.

Reviewer 2: Can we take the opportunity to fix the sentence/grammar here?

I think it reads better as:

>   When using Deadline 10.1.19 or higher, RFDK enables [Deadline Secrets Management](https://docs.thinkboxsoftware.com/products/deadline/10.1/1_User%20Manual/manual/secrets-management/deadline-secrets-management.html) on the `Repository` by default. RFDK is able to ...

Also, for:

>   ... and you cannot accidentally delete the credentials that the database is configured to use.

I understand that you're referring to deletion via:

1.  CDK/Cloudformation deployments/destroys and
2.  when using [tiered architecture](https://docs.aws.amazon.com/rfdk/latest/guide/best-practices.html#organizing-your-stacks-multi-tiered-architecture) where the `Repository` is in a dependent stack of the database.

but I'm struggling to find the appropriate clause to qualify this statement since users can still "accidentally delete the credentials that the database is configured to use" outside of CloudFormation. Perhaps we just remove the clause entirely and just have it as:

>   When RFDK generates the admin credentials for you, it puts them in the same Stack as your database so that their lifecycles are in sync. If your database is imported ...

What do you think?

Author: Updated based on yours and Eugene's comments. Let me know what you think.

Reviewer 2: Much better! thanks

> when using Deadline 10.1.19 or higher. RFDK is able to create administator credentials for Deadline Secrets Management, store them in AWS Secrets Manager, and configure the Deadline Repository with those credentials. When RFDK generates the admin credentials for you, it puts them in the same Stack as your database so that their lifecycles are in sync and you cannot accidentally delete the credentials that the database is configured to use. If your database is imported into your RFDK app rather than being generated by it, RFDK will not allow your app to have credentials generated automatically and you'll need to create your own AWS Secret with the credentials you'd like to use and import it.

Reviewer 3: ```suggestion
when using Deadline 10.1.19 or higher. RFDK is able to create administrator credentials for Deadline Secrets Management, store them in AWS Secrets Manager, and configure the Deadline Repository with those credentials. When RFDK generates the admin credentials for you, it puts the `Secret` in the same `Stack` as your database so that their lifecycles are in sync, which ensures that RFDK does not destroy the `Secret` with your admin credentials while the database is still in use. If your database is imported into your RFDK app rather than being generated by it, RFDK will not allow your app to have credentials generated automatically and you'll need to import your own AWS Secret with the credentials you'd like to use.
```

Couple things:
1. Looks like there's a typo in "administator" that I missed from the previous PR 😞 could you fix it?
2. I think we can be reword the reason we tie the lifecycle of the Secret to the database so it's more clear that we prevent the CDK deployment from destroying the secret while it's still required by the DB

> when using Deadline 10.1.19 or higher. RFDK is able to create administrator credentials for Deadline Secrets Management, store them in AWS Secrets Manager, and configure the Deadline Repository with those credentials. When RFDK generates the admin credentials for you, it puts the `Secret` in the same `Stack` as your database so that their lifecycles are in sync, which ensures that RFDK does not destroy the `Secret` with your admin credentials while the database is still in use. If your database is imported into your RFDK app rather than being generated by it, RFDK will not allow your app to have credentials generated automatically and you'll need to import your own AWS Secret with the credentials you'd like to use.

Reviewer: I believe it is worth to mention that this Secret by default has `RETAIN` remove policy and will not be removed with stack, so customer have to remove it manually after destroying farm.

});
```

> Once the repository has been deployed and the admin credentials are configured, changing them in the Secret and redeploying your RFDK app will not change the credentials the deployed Deadline will use. For more information about updating admin credentials and other considerations for using Deadline Secrets Management in RFDK, please see the [RFDK Developer Guide](https://docs.aws.amazon.com/rfdk/latest/guide/deadline-secrets-management-rfdk.html).

Reviewer 2: Can we change this to link to the proper credential rotation procedure and warn about the wrong way after?

Author: Updated

});
```

> Once the repository has been deployed and the admin credentials are configured, changing them in the Secret and redeploying your RFDK app will not change the credentials the deployed Deadline will use. If you need to change the credentials, you will first need to manually change them in Deadline before updating the Secret, or else RFDK will no longer be able to automate any administrator operations during re-deployments. For more information about properly updating admin credentials and other considerations for using Deadline Secrets Management in RFDK, please see the [RFDK Developer Guide](https://docs.aws.amazon.com/rfdk/latest/guide/deadline-secrets-management-rfdk.html#rotating-deadline-secrets-management-administrator-credentials).

Reviewer 2: Hmm.... how do you feel about flipping the order here?

```suggestion
Please consult the RFDK developer guide topic on [using Deadline Secrets Management in RFDK](https://docs.aws.amazon.com/rfdk/latest/guide/deadline-secrets-management-rfdk.html#using-deadline-secrets-management-in-rfdk). Please note that there is a specific process for [rotating your administrator credentials](https://docs.aws.amazon.com/rfdk/latest/guide/deadline-secrets-management-rfdk.html#rotating-deadline-secrets-management-administrator-credentials).

**WARNING:** Once the repository has been deployed and the administrator credentials are configured, changing the credential values in the Secret and re-deploying your RFDK app will not update the credentials in the Deadline Repository. You must first change the credentials in the Deadline Repository using the instructions linked above. Doing this incorrectly may cause the RFDK Render Queue to stop functioning. If you have done this and cannot recover the previous credentials, you will need to follow instructions for [forgotten Deadline Secrets Management administrator credentials](https://docs.aws.amazon.com/rfdk/latest/guide/deadline-secrets-management-rfdk.html#forgotten-deadline-secrets-management-administrator-credentials).
```

Author: Updated
